### PR TITLE
support float8 weight caching for gradient accumulation/PP

### PR DESCRIPTION
**Pull request by vkuzo**
Summary:

In the cases where the optimizer update does not happen after every forward such as microbatching/PP, we can save the casted weight to trade some time for memory.

For now I'm just testing out performance+accuracy. We can improve on the API in future PRs.  The current code is torch.compile friendly which is nice.

In terms of accuracy this should be no change, I will validate this further if we want to land this.

For performance, on @drisspg's LLaMa 7B pretrain script, with bsz==128 and micro_bsz == 1:

1. baseline bf16 + compile: 2.38 it/s
2. delayed scaling + compile: 2.80 it/s (1.18x over baseline)
3. delayed scaling + compile + this PR: 3.04 it/s (1.28x over baseline)

Test Plan:

```
pytest test/test_base.py -s -k test_weight_caching
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: float8_experimental/float8_tensor.py [hunk 1]
```diff
@@ -123,6 +132,9 @@ def __tensor_flatten__(self):
 
     @staticmethod
     def __tensor_unflatten__(inner_tensors: Dict, metadata):
+        # TODO(TBD): this seems unused, and it's out of date after
```
Comment (drisspg) [inline float8_experimental/float8_tensor.py:135]: I was looking into this today and I do think that this is used it is just that in full graph compile land we get away with this not being a problem we should update this to the new signature

Comment (vkuzo) [reply float8_experimental/float8_tensor.py:135]: hmm, do you have more context?

If I understand the current state correctly:
1. the currently checked in code is broken because the signature of __tensor_unflatten__ is no longer valid after https://github.com/pytorch/pytorch/pull/114311
2. if this code path is exercised by dynamo, it will crash because of (1)
3. all tests pass, so I'm guessing this is dead code at the moment

Comment (drisspg) [reply float8_experimental/float8_tensor.py:135]: @bdhirsh I imagine that although this is broken we are not encountering this because we can full graph compile and the tensor subclass vanishes from aot, is that correct or am I missing something? So although this is not exercised today if we decided to expose float8tensor subclass as more of a first ux and they entered compiled functions we would need this again

Comment (bdhirsh) [reply float8_experimental/float8_tensor.py:135]: Yep - I think that if your subclass is just a temporary that poofs during tracing (so not a graph input or output), then we won't end up hitting the flatten/unflatten calls at all. We'll definitely need it though if/when the subclass ever becomes a graph input/output.

This can happen if you make the subclass a module parameter directly, **but** it could also just happen via graph breaks, if there happens to be a graph break after we create the float8 tensor but before it vanishes (which to be fair would probably be bad for the level of perf that float8 wants anyway).

Comment (vkuzo) [reply float8_experimental/float8_tensor.py:135]: thanks @bdhirsh  for explaining.  I'll remove the TODO from this PR, and looks like we should fix the signature and add a unit test with the conditions you described.

Comment (vkuzo) [reply float8_experimental/float8_tensor.py:135]: put up https://github.com/pytorch-labs/float8_experimental/pull/166, it doesn't work yet, @bdhirsh  any thoughts on the error message in that PR?

Comment (vkuzo) [reply float8_experimental/float8_tensor.py:135]: yep its in the test plan (not in CI)!

Comment (bdhirsh) [reply float8_experimental/float8_tensor.py:135]: missed it the first time https://gist.github.com/vkuzo/ba98a01a459fb9c966f167d8ecca1780#file-gistfile0-txt-L137.

Hmm, this looks like the same error @drisspg and I noticed a month or two ago, where (somehow) we end up with multiple fake modes floating around. This probably happens somewhere in dynamo, since dynamo is responsible for creating the FakeTensorMode.

This bug might be pretty involved, so just checking - "get float8tensor working with graph breaks" a blocker? (Regardless I agree that we should make it work)

File: test/test_base.py [hunk 2]
```diff
@@ -231,6 +234,36 @@ def test_type_cast(self, linear_type: LinearType, linear_dtype: torch.dtype):
...
+        if use_compile:
+            m = torch.compile(m)
+
+        config.weight_cache_enabled = False
+
+        y1 = m(x)
+        y1.sum().backward()
+        grad1 = m.weight.grad.clone().detach()
+
+        config.weight_cache_enabled = True
```
Comment (drisspg) [inline test/test_base.py:256]: I was also thinking about this after you brought this up: https://github.com/pytorch-labs/float8_experimental/commit/c09b67ffee21ffd1ac361ce9c0114e8eadd852bb

was more curious as to what the ideal ux would be for turning on and off the weight-cachihng

Comment (vkuzo) [reply test/test_base.py:256]: I think right now it has to be "what's the UX which isn't terribly slow after torch.compile".  I tried a couple of things and the global variable was the only one which didn't have a large overhead :(

TE requires user to explicitly pass a `is_first_microbatch` flag, I didn't do that because that requires a model rewrite.  The global thing allows the signature of `Float8Linear`'s forward to stay the same.

File: float8_experimental/float8_linear.py [hunk 3]
```diff
@@ -148,6 +153,15 @@ def __init__(self, *args, **kwargs):
         # will access the scale when it has ensured that it is on GPU.
         self._float8_tensor_ctor = lambda *args, **kwargs: Float8Tensor(*args, **kwargs)
 
+        if config.allocate_float8_weight_cache_buffers:
+            # this is a buffer to get `to(dtype)` for free
+            # TODO(future): hide this from serialization
+            # TODO(future): force this to stay in float8_e4m3fn
+            self.register_buffer(
```
Comment (drisspg) [inline float8_experimental/float8_linear.py:160]: https://github.com/pytorch-labs/float8_experimental/blob/main/float8_experimental/float8_linear.py#L175

Could probably use something like this for the second todo

File: float8_experimental/config.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+
+# If True, allocates buffers for float8 weight cache
+allocate_float8_weight_cache_buffers = False
```
Comment (drisspg) [inline float8_experimental/config.py:8]: I think some more comments on how users are expected to use this in code

File: float8_experimental/float8_linear.py [hunk 5]
```diff
@@ -204,8 +218,30 @@ def cast_w_to_float8(
             torch.float8_e4m3fn,
             is_amax_initialized,
         )
+
+        if config.weight_cache_enabled:
+            assert config.allocate_float8_weight_cache_buffers
```
Comment (drisspg) [inline float8_experimental/float8_linear.py:223]: Also maybe a better error message, "if you are using weight caching you need to enable allocate_float8_weight_cache_buffers = False before performing module swap or construction of Float8Linear module"

**Review by drisspg — APPROVED** [review]
Left some comments, mostly around documentation but seems good

File: float8_experimental/config.py (hunk 4 above)
Comment (vkuzo) [reply float8_experimental/config.py:12]: added to readme

File: float8_experimental/float8_linear.py (hunk 3 above)
Comment (vkuzo) [reply float8_experimental/float8_linear.py:184]: yep, we can fix that in a future PR!

File: float8_experimental/float8_tensor.py [hunk 6]
```diff
@@ -122,7 +131,7 @@ def __tensor_flatten__(self):
         return ["_data", "_scale"], ctx
 
     @staticmethod
-    def __tensor_unflatten__(inner_tensors: Dict, metadata):
+    def __tensor_unflatten__(inner_tensors: Dict, metadata, outer_size, outer_stride):
```
Comment (vkuzo) [inline float8_experimental/float8_tensor.py:134]: not related to this PR, but fixing it to adhere to the changes in https://github.com/pytorch/pytorch/pull/114311/files

File: float8_experimental/float8_tensor.py (hunk 1 above)
Comment (vkuzo) [reply float8_experimental/float8_tensor.py]: not technically a blocker, but would be really nice for it to work
